content/300-guides/100-performance-and-optimization/090-metrics.mdx

```diff
@@ -0,0 +1,416 @@
+
+Use version `3.15.0` or higher of the `prisma` and `@prisma/client` npm packages.
+
+```terminal
+npm install prisma@latest --save-dev
+npm install @prisma/client@latest --save
+```
+
+### Step 2: Enable the feature flag in the Prisma schema file
+
+In the `generator` block of your `schema.prisma` file, enable the `metrics` feature flag:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["metrics"]
+}
+```
+
+## Retrieve metrics in JSON format
+
+When you retrieve metrics in JSON format, you can use them in the format they are returned, or [send them to StatSD](#use-prisma-metrics-with-statsd) to visualize how they change over time.
+
+To retrieve metrics in JSON format, add the following lines to your application code:
+
+```ts
+let metrics = await prisma.$metrics.json()
+console.log(metrics)
+```
+
+This returns metrics as follows:
+
+```json
+{
+  "counters": [
+    {
+      "key": "query_total_operations",
+      "labels": {},
+      "value": 0,
+      "description": "The total number of operations executed"
+    },
+    {
+      "key": "query_total_queries",
+      "labels": {},
+      "value": 0,
+      "description": "The total number of queries executed"
+    }
+  ],
+  "gauges": [
+    {
+      "key": "pool_active_connections",
+      "labels": {},
+      "value": 0,
+      "description": "The number of active connections in use"
+    },
+    {
+      "key": "pool_idle_connections",
+      "labels": {},
+      "value": 21,
+      "description": "The number of connections that are not being used"
+    },
+    {
+      "key": "pool_wait_count",
+      "labels": {},
+      "value": 0,
+      "description": "The number of workers waiting for a connection"
+    },
+    {
+      "key": "query_active_transactions",
+      "labels": {},
+      "value": 0,
+      "description": "The number of active transactions"
+    }
+  ],
+  "histograms": [
+    {
+      "key": "pool_wait_duration_ms",
+      "labels": {},
+      "value": {
+        "buckets": [
+          [0, 0],
+          [1, 0],
+          [5, 0],
+          [10, 0],
+          [50, 0],
+          [100, 0],
+          [500, 0],
+          [1000, 0],
+          [5000, 0],
+          [50000, 0]
+        ],
+        "sum": 0,
+        "count": 0
+      },
+      "description": "The wait time for a worker to get a connection."
+    }
+  ]
+}
+```
+
+### Use Prisma metrics with StatsD
+
+You can send JSON-formatted metrics to StatsD to visualize your metrics data over time.
+
+<Admonition type="info">
+  Note: When you send data to StatsD, you must send both Prisma counter and
+  gauge metrics as gauge data. This is because StatsD expects counter metrics to
+  be a series of incremented or decremented values, but Prisma counter metrics
+  return actual values.
+</Admonition>
+
+In the following example, we send metrics to StatsD every 10 seconds. This timing aligns with the default 10s flush rate of StatsD.
```

Comment:
This code would be part of my application and just run endlessly to get the metrics every 10 seconds and transfer them to statsd?

Reply:
Yes. Not quite the usual way to work StatsD to be honest, but prometheus and statsd have very different quite incompatible ways of operations, so I guess one of them is bound to look suboptimal.